debugger/debuggers.go

```diff
@@ -0,0 +1,54 @@
+package debugger
+
+import (
+	"fmt"
+	"github.com/sstallion/go-hid"
+)
+
+var DebuggerMap = make(map[uint32]func() (Debugger, error))
```

Comment:
Okay, I see what's going on now.  You want to dynamically detect the available debuggers and return whatever is present as an implementation-specific thing (ICE, J-Link, etc).  This is certainly one way to do it.  But my worry here is the less specific each target is about it's expected environment, the more weird gotchas we're going to run into.

As an example: Let's say we have two CMSIS-DAP compatible debuggers, an Atmel ICE and something we'll just call X.  And the ICE supports the DAP_ResetTarget command, but the X does not.  Maybe the X can still reset the target via register writes to the target, but that procedure is different.  Think about how this is going to affect the target code...

If we want to be able to issue a command like `goocd -target=atsame51-cmsisdap -reset` which will cause a reset command to be issue.  In this case we would be assuming that a generic "reset" can be somehow converted into these two different code paths - one using DAP_ResetTarget and the other using register writes to cause a software reset.  This is going to be a fair bit of complexity and abstraction (and if you think it's worth doing it this way, then it would be a good idea to get that code written out so we can see exactly what effects that decision makes in the code structure).  Contrast it with having two commands:

`goocd -target=atsame51-atmelice -reset`

`goocd -target=atsame51-x -reset` (`x` would be whatever actual name of the debugger device)

In this case, the target code has this difference baked into it, roughly as psuedocode it would be:

```
// atsame51-atmelice
if reset {
  debugger.DAP_ResetTarget
}
```

```
// atsame51-x
if reset {
  debugger.WriteReg32(same51.RESET_REG, same51.RESET_VAL)
}
```

Having this code difference be part of the target I think is an important structural distinction.  It means more code duplication (every chip+debugger combination needs a separate target), but the reward is that each such configuration is significantly simpler and easier to understand and debug. 

(To further clarify: If you look at the -x version above, you'll notice that this code needs to know that CMSIS dap doesn't DAP_ResetTarget, and that WriteReg32 is available, and which exact registers for this chip are needed for the reset.  I think you'll find that if you moved this anywhere else you'll have a hard time finding any other place where it is appropriate to know all of these details - which is going to lead to some sort of abstraction such as an interface or specialized struct fields, etc. and all of the complexity that comes with that - for EACH issue where this arises, resetting targets will certainly not be the last thing like this but rather the first of many.)

We can discuss in more detail, but I think this is a critical code structure issue to resolve and agree on.